batch/batch/server/database.py

```diff
@@ -0,0 +1,167 @@
+
+    async def create_temporary_table(self, name, schema, keys):
+        name = await self.temp_table_name(name)
+        t = await Table(self, name, schema, keys)
+        return t
+
+    def create_temporary_table_sync(self, name, schema, keys):
+        return run_synchronous(self.create_temporary_table(name, schema, keys))
+
+
+def make_where_statement(items):
+    template = []
+    values = []
+    for k, v in items.items():
+        if isinstance(v, list):
+            if len(v) == 0:
+                template.append("FALSE")
+            else:
+                template.append(f'`{k.replace("`", "``")}` IN %s')
+                values.append(v)
+        else:
+            template.append(f'`{k.replace("`", "``")}` = %s')
+            values.append(v)
+
+    template = " AND ".join(template)
+    return template, values
+
+
+@asyncinit
+class Table:  # pylint: disable=R0903
+    async def __init__(self, db, name, schema, keys):
+        self.name = name
+        self._db = db
+        await self._db.create_table(name, schema, keys)
+
+    async def _new_record(self, items):
+        names = ", ".join([f'`{name.replace("`", "``")}`' for name in items.keys()])
+        values_template = ", ".join(["%s" for _ in items.values()])
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                sql = f"INSERT INTO `{self.name}` ({names}) VALUES ({values_template})"
+                await cursor.execute(sql, tuple(items.values()))
+                id = cursor.lastrowid  # This returns 0 unless an autoincrement field is in the table
+        return id
+
+    async def _update_record(self, where_items, set_items):
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                if len(set_items) != 0:
+                    where_template, where_values = make_where_statement(where_items)
+                    set_template = ", ".join([f'`{k.replace("`", "``")}` = %s' for k, v in set_items.items()])
+                    set_values = set_items.values()
+                    sql = f"UPDATE `{self.name}` SET {set_template} WHERE {where_template}"
+                    await cursor.execute(sql, (*set_values, *where_values))
+
+    async def _get_records(self, where_items, select_fields=None):
+        assert select_fields is None or len(select_fields) != 0
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                where_template, where_values = make_where_statement(where_items)
+                select_fields = ",".join(select_fields) if select_fields is not None else "*"
+                sql = f"SELECT {select_fields} FROM `{self.name}` WHERE {where_template}"
+                await cursor.execute(sql, where_values)
+                result = await cursor.fetchall()
+        return result
+
+    async def _has_record(self, where_items):
```

Comment:
missing test

Reply:
It's implicitly in there for create_drop. Let me know if that's not good enough.

Reply:
I actually don't see where Table's `_has_record` method is used?

Reply:
Added. Got confused with has_table